File: MiniGame-Java/src/main/java/rs/lab/mges/game/Powerup.java
package rs.lab.mges.game;

import io.github.libsdl4j.api.render.SDL_Texture;
import rs.lab.mges.engine.Drawable;
import rs.lab.mges.engine.Drawable.Sprite;
import rs.lab.mges.engine.SDLUtils;
import rs.lab.mges.engine.SDLUtils.Vector2f;

import static io.github.libsdl4j.api.sensor.SdlSensorConst.SDL_STANDARD_GRAVITY;

public abstract class Powerup extends Sprite {

    public boolean active;
    public int point;

    private float waitTimer = 0f;

    public Powerup(SDL_Texture frame, int point) {
        super(frame, 32, 32);
        this.point = point;
        hitbox = SDLUtils.rect(0, 10, size.x, size.y - 10);
    }

    public boolean hit(Hero player) {
        if (Drawable.AABB(getHitbox(), player.getHitbox())) {
            active = false;
            return true;
        }

        return false;
    }

    @Override
    public void update(float dt) {
        if (position.y < 280) {
            velocity.y += SDL_STANDARD_GRAVITY;
            position.y += dt * velocity.y;
        } else {
            position.y = 280f;
            waitTimer += dt;
            if (waitTimer > 3.0) {
                if (alpha < 10) {
                    alpha = 0;
                    active = false;
                } else {
                    alpha -= 5;
                }
            }
        }
    }

    public void reset(float x) {
        position = new Vector2f(x, 100.0f);
        velocity = new Vector2f(0.0f, 50.0f);
        waitTimer = 0.0f;
        alpha = 255;
        active = true;
        oldPosition = null;
    }

    public static class Bomb extends Powerup {

        public Bomb(RabbitGame game) {
            super(game.assets.bomb, 500);
        }
    }

    public static class Medkit extends Powerup {

        public Medkit(RabbitGame game) {
            super(game.assets.medKit, 500);
        }
    }
}

File: MiniGame-Python/src/game/Powerup.py

from engine import Sprite
from sdl2 import * # type: ignore

from engine.Utils import Vector2f
from game import Hero, RabbitGame

__all__ = ['Bomb', 'Medkit']

class Powerup(Sprite):
    def __init__(self, frame, points):
        super().__init__(frame, 32,32)
        self._active = False
        self._points = points
        self._waitTimer = 0.0
        self._hitbox = SDL_Rect(0,10, int(self._size.x), int(self._size.y-10))

    def hit(self, hero: Hero):
        if Sprite.AABB(self.hitbox, hero.hitbox):
            self._active = False
            return True
        else:
            return False
        
    def update(self, dt):
        if self._position.y < 280.0:
            self._velocity.y += SDL_STANDARD_GRAVITY
            self._position.y += dt * self._velocity.y
        else:
            self._position.y = 280.0
            self._waitTimer += dt
            if self._waitTimer > 3.0:
                if self._alpha < 10:
                    self._alpha = 0
                    self._active = False
                else:
                    self._alpha -= 5
    
    def reset(self, x:float):
        self._position = Vector2f(x, 100.0)
        self._velocity = Vector2f(0.0, 50.0)
        self._waitTimer = 0.0
        self._alpha = 255
        self._active = True
        self._oldPosition = None

class Bomb(Powerup):
    def __init__(self, game: RabbitGame):
        super().__init__(game.assets.bomb, 500)

class Medkit(Powerup):
    def __init__(self, game: RabbitGame):
        super().__init__(game.assets.medKit, 500)
